Add warning for outside of project folder files

blurymind opened the pull request:
This adds  a warning for resources that are residing outside of the project folder- so issues like this:
https://github.com/4ian/GDevelop/issues/1048

are prevented

Screenshots
![warningPath](https://user-images.githubusercontent.com/6495061/57954317-74df9180-78ea-11e9-9174-5cf8a358c248.PNG)
![warningBetterMessage](https://user-images.githubusercontent.com/6495061/57957001-c5f38380-78f2-11e9-862b-a77ee5bc9018.PNG)

4ian commented:
Nice! Let me check that today :)

4ian commented on newIDE/app/src/ResourcesList/index.js:
```diff
@@ -31,7 +31,7 @@ const styles = {
   renamedResource: ?gdResource,
   searchText: string,
-  resourcesWithMissingPath: { [string]: boolean },
+  resourcesWithErrors: { [string]: 0 | 1 | 2 },
```
While it's a good idea to use an "enum" like this, it would be clearer I think with strings:

`resourcesWithErrors: { [string]: "no-error" | "not-found" | "outside-project-foler" }`

You'll have to change the rest of the references to 0, 1 and 2, but this should be clearer :)

4ian commented on newIDE/app/src/UI/SortableVirtualizedItemList/ItemRow.js:
```diff
@@ -28,7 +28,7 @@ type Props = {
   onItemSelected: () => void,
   buildMenuTemplate: Item => any,
-  hasErrors: boolean,
+  errorStatus: 0 | 1 | 2,
```
Here I think it would be better to have `errorsStatus: '' | 'error' | 'warning`.

blurymind replied:
I was thinking the same thing, but wasnt sure. Will do the changes tonight :)

4ian commented on newIDE/app/src/ResourcesList/LocalResourceSources.js:
```diff
@@ -164,6 +165,8 @@ const selectLocalResourcePath = (
       },
       paths => {
         if (!paths) return resolve([]);
+        if (!userWantsToUseExternalPath(project, paths[0])) return resolve([]);
```
This looks like the correct place where to put this check 👍

4ian commented on newIDE/app/src/ResourcesList/ResourceUtils.js:
```diff
@@ -48,12 +49,42 @@ export const getLocalResourceFullPath = (
 };
 
-export const resourceHasValidPath = (
+export const isResourcePathinProjectFolder = (
```
nitpicking: `isPathInProjectFolder` (actually, you check the path, you don't care about the fact it's a path used for a resource)

4ian commented on newIDE/app/src/ResourcesList/ResourceUtils.js:
```diff
@@ -48,12 +49,42 @@ export const getLocalResourceFullPath = (
+  return resourcePath.includes(projectPath);
+};
+
+export const userWantsToUseExternalPath = (
```
nitpicking: `userWantsToUseExternalPath` looks like it's something that return true or false, but it's actually doing some more complex stuff (i.e: ask something to the user).
Also, I find it unusual to include "user" in the name of the method. I prefer method starting with a verb that describe what is being done. I would suggest: `confirmResourcePath`

confirm hints at the usage of `confirm` method (so a user input will happen). The name is more generic (you could say too generic), but this means that we can later add more validation here.

4ian commented:
Looks mostly good :) I've changed the message and added comments about the typing of the error/warnings, and naming of methods. Apart from that, I think we're having something good and useful here.

blurymind commented:
all done :)

blurymind commented:
ah sec, I broke the flow with a master merge

4ian commented on newIDE/app/src/UI/SortableVirtualizedItemList/index.js:
```diff
@@ -96,10 +96,10 @@ class ItemsList extends React.Component<ItemsListProps, *> {
               getThumbnail={getThumbnail ? () => getThumbnail(item) : undefined}
               selected={item === selectedItem}
               onItemSelected={this.props.onItemSelected}
+              errorStatus={erroredItems ? erroredItems[item.getName()] : 0}
```
`erroredItems ? erroredItems[item.getName()] : ''` if I'm not mistaken

blurymind replied:
thanks :D sorry I missed it

4ian replied:
Also, if `erroredItems` does not contain `item.getName()`, you'll get undefined.
To be 100% correct, you should do:
`erroredItems ? erroredItems[item.getName()] || '' : ''}`

(which I admit, start to be a bit cryptic, but when you're used to it, this translate to "no error if erroredItems does not exist, and no error if the item is not found in these erroredItems".)

blurymind replied:
good catch! 👍 added to the pr. Sorry I missed it :)

4ian commented:
Almost there ;) Can you add images to multiple type of objects to be sure this is working properly with sprite, tiled sprite, panel sprite, etc but also audio (in events) and videos? 
No reason this would be failing as all are using the same method, but as it's a super used feature I want to be extra sure we don't have any weird behavior here :)

blurymind commented:
@4ian hope you don't mind that I added a command to run electron to package.json, I kept forgetting the command to run it, so I added 
`npm run electron-win`
it's easier to remember

4ian commented:
Mmm, ideally would have been cool to do two pull requests for that. Having a single one means that now I can't nitpick on the commands without holding the merge of the rest :)
As possible, prefer starting a new branch and a new PR, even for small changes like this.
Also when I merge the PR into master, I "squash" the commits with GitHub, so that all the commits are merged into a single one (to keep a clean git history). If we have multiple modifications, this make it harder in the future to browse the history and isolate changes.

Would you be able to push a commit reverting the changes? You can actually keep your changes in git by starting a new branch (let's call it "electron-commands"), then going back to your original branch, add a commit to revert the changes and push. Then go back to "electron-commands" and push into another PR.
Ideally (but this requires more mastery of git), you can print the commit hash (using `git log`), create a new branch from `master`, then use `git cherry-pick <hash of your commit>` to copy your commit containing the new commands into this new branch.

blurymind commented:
Ah sorry, I will split them up into separate pulls tonight :)

4ian commented on newIDE/app/src/UI/SortableVirtualizedItemList/ItemRow.js:
```diff
@@ -27,8 +27,9 @@ type Props = {
   getThumbnail?: () => string,
   selected: true,
   onItemSelected: () => void,
+  buildMenuTemplate: Item => any,
```
This seems to be not related to this PR, and probably and copy-paste of an old version of the file. Can you put `buildMenuTemplate: () => any,` back again?

blurymind replied:
i just noticed that I have it twice :o looks like bad resolve of conflicts. Fixed it now

4ian commented:
Looks mostly ok! Automatic tests are having a hard time as I've pushed things to master too, but everything should settle down in a bit of time.
Check my last comment, rest looks good!

blurymind commented:
thank you for  the patience  
Also for the awesome updates you've been working on! :D

4ian commented:
I've merged this but there was a slight bug. If the ressource is missing, but located outside the project folder (for example, if the file is "../file-that-does-not-exist.png"), then it is shown with a warning instead of an error. I've fixed this on master.

blurymind commented:
> I've merged this but there was a slight bug. If the ressource is missing, but located outside the project folder (for example, if the file is "../file-that-does-not-exist.png"), then it is shown with a warning instead of an error. I've fixed this on master.

good catch! Thank you for fixing 😸
